gt/main.c

```diff
@@ -131,36 +129,78 @@ gt_parse_incoming_pkt(struct rte_mbuf *pkt, struct gt_packet_headers *info)
 		info->inner_ip_ver = ETHER_TYPE_IPv4;
 		info->l4_proto = inner_ipv4_hdr->next_proto_id;
 		info->l4_hdr = &inner_ipv4_hdr[1];
+
+		if (rte_ipv4_frag_pkt_is_fragmented(inner_ipv4_hdr)) {
+			info->frag = true;
+			info->l2_len = parsed_len;
+			info->l3_len = sizeof(*inner_ipv4_hdr);
+		}
 	} else if (likely(inner_ip_ver == 6)) {
-		/*
-	 	 * Make sure that the packet has space for
-		 * at least 4 bytes for the l4 header.
-	 	 */
-		if (pkt->data_len < parsed_len + sizeof(struct ipv6_hdr) + 4)
+		if (pkt->data_len < parsed_len + sizeof(struct ipv6_hdr))
 			return -1;
 
 		inner_ipv6_hdr = (struct ipv6_hdr *)info->inner_l3_hdr;
 		info->inner_ip_ver = ETHER_TYPE_IPv6;
 		info->l4_proto = inner_ipv6_hdr->proto;
 		info->l4_hdr = &inner_ipv6_hdr[1];
+
+		info->frag_hdr =
+			rte_ipv6_frag_get_ipv6_fragment_header(inner_ipv6_hdr);
+
+		if (info->frag_hdr != NULL) {
+			info->frag = true;
+			info->l2_len = parsed_len;
+			info->l3_len = sizeof(*inner_ipv6_hdr) +
+				sizeof(*info->frag_hdr);
+		}
 	} else
 		return -1;
 
 	return 0;
 }
 
+static struct rte_mbuf *
+gt_reassemble_incoming_pkt(struct rte_mbuf *pkt, uint64_t tms,
+	struct gt_packet_headers *info, struct gt_instance *instance)
+{
+	/* Prepare mbuf: setup l2_len/l3_len. */
+	pkt->l2_len = info->l2_len;
+	pkt->l3_len = info->l3_len;
+
+	if (info->inner_ip_ver == 4) {
+		/* Process this IPv4 fragment. */
+		return rte_ipv4_frag_reassemble_packet(
+			instance->frag_tbl, &instance->death_row,
+			pkt, tms, info->inner_l3_hdr);
```

Comment:
I think the last parameter should be the pointer to the IPv4 header that has been marked as fragmented, which is the outer header.